com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl

```diff
@@ -343,13 +346,38 @@ half AdditionalLightRealtimeShadow(int lightIndex, float3 positionWS, half3 ligh
     return SampleShadowmap(TEXTURE2D_ARGS(_AdditionalLightsShadowmapTexture, sampler_AdditionalLightsShadowmapTexture), shadowCoord, shadowSamplingData, shadowParams, true);
 }
 
-half GetShadowFade(float3 positionWS)
+half GetMainLightShadowFade(float3 positionWS)
+{
+    // This code computes distance between capsule and point.
+    // Finally distance is used with shadow fade scale and bias to compute shadow fade blend value.
+
+    float3 shadowFadeCenter = _CascadeShadowSplitSphereLast.xyz;
+    float3 worldSpaceCameraOppositeDir = -UNITY_MATRIX_V[2].xyz; // Move it to other property for faster access?
+
+    // Here we treat problem as finding shortest distance between line segment and point.
+
+    float3 lineDir = worldSpaceCameraOppositeDir;
+    float3 lineCenter = shadowFadeCenter;
+
+    float3 lineToPoint = positionWS - lineCenter;
+
+    // Find shortest vector between line and point
+    float distanceFromLineCenter = min(dot(lineDir, lineToPoint), 0);
+    float3 offsetFromLineCenter = lineDir * distanceFromLineCenter;
+    float3 shortestLineToPoint = lineToPoint - offsetFromLineCenter;
+
+    float distanceSq = dot(shortestLineToPoint, shortestLineToPoint);
+    half fade = saturate(distanceSq * _MainLightShadowParams.z + _MainLightShadowParams.w);
+    return fade;
```

Comment:
This needs to be the min between the edge of the sphere, the shadow distance, and the far plane for the cascade.

Reply:
the last cascade should already be the minimum of shadow distance, far plane and of course sphere (As it is). Maybe I missunderstood here?

Reply:
Unfortunately it is a false assumption. Stable fit was never implemented properly which gives this false impression. The implementation does try to do this, but it doesn't work, though the cascade frustum should actually be entirely within each sphere. 

![image](https://user-images.githubusercontent.com/72749964/105208982-99711f80-5b49-11eb-81c0-cbd4c4502bae.png)

Reply:
The assumption is only true for square aspect ratios.

Reply:
Mhh, is it in all cases? As I did not noticed visually this effect. If it is rare case and will be fixed with ure internal change, maybe we can leave it as it is. Also HDRP should have same issue, as they also do shadow fade from cascade.

Reply:
I don't think it is a rare issue and it should be very common. These types of shadow problems are creating a lot of frustrations with users. With my change, this problem wouldn't happen for all cascades, so this fix you have made would not be necessary, but @phi-lira would like us to keep this functionality as a performance/quality optimization (and the default behavior still needs to be discussed). Yeah, HDRP has the same issue, but we should definitely fix it for URP. HDRP use the cascade sphere for a few different things.

Reply:
After offline discusion with Nigle we decided to revert shadow fade from last cascade. As this will be fixed by Nigle stable fit and current fix does not handle case described in here. HDRP is also suffering from that, so I think we should not follow their path here.
![hdrp-issue](https://user-images.githubusercontent.com/24258017/105822705-0c085200-5fc5-11eb-8351-bffd772ceede.gif)